social_media_publishers/notification.py

```diff
@@ -0,0 +1,89 @@
+from datetime import datetime
+from social_media_publishers.publisher import Publisher
+from firestore.incidents import Incident
+from firestore.tokens import Token
+from datetime import datetime, timedelta
+import concurrent.futures
+
+from exponent_server_sdk import (
+    DeviceNotRegisteredError,
+    PushClient,
+    PushMessage,
+    PushServerError,
+)
+
+
+class PushNotification(Publisher):
+    def __init__(self) -> None:
+        self.push_client = PushClient()
+
+    def publish(self, incident: Incident) -> datetime:
+        # Fetch tokens in batches
+        batch_size = 10
+        batch_no = 0
+        tokens = Token.collection.fetch(batch_size)
+
+        while True:
+            batch_no += 1
+            print(f"Grouping batch {batch_no} ...")
+
+            # Group tokens by first two characters
+            grouped_tokens = {}
+            for token in tokens:
+                group_key = token.token[18:20]
+                # example token: ExponentPushToken[hty9r4BGScZt5gloxXCITM]
+                if group_key not in grouped_tokens:
+                    grouped_tokens[group_key] = []
+                grouped_tokens[group_key].append(token)
+
+            if not grouped_tokens:
+                break
+
+            # Process each group in parallel
+            with concurrent.futures.ThreadPoolExecutor() as executor:
```

Comment:
pls check Google App engine see if multi-threading is supported or not.

Reply:
multi-threading is not supported on Google App Engine Standard. However, we can use google-cloud Task Queues which is a way of distributing work across many concurrent requests